sound/soc/sof/ipc4-topology.c

```diff
@@ -1976,12 +1978,30 @@ static int sof_ipc4_control_setup(struct snd_sof_dev *sdev, struct snd_sof_contr
 	return 0;
 }
 
+static int sof_ipc4_pipeline_send_kcps(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
+				       __s32 adjust)
+{
+	const struct sof_ipc_ops *iops = sdev->ipc->ops;
+	struct sof_ipc4_msg msg = {
+		.primary = SOF_IPC4_MSG_TARGET(SOF_IPC4_MODULE_MSG) |
+			SOF_IPC4_MSG_DIR(SOF_IPC4_MSG_REQUEST) |
+			SOF_IPC4_MOD_ID(SOF_IPC4_MOD_INIT_BASEFW_MOD_ID) |
+			SOF_IPC4_MOD_INSTANCE(SOF_IPC4_MOD_INIT_BASEFW_INSTANCE_ID),
+		.extension = SOF_IPC4_MOD_EXT_MSG_PARAM_ID(SOF_IPC4_FW_PARAM_REGISTER_KCPS),
+		.data_size = sizeof(adjust),
+		.data_ptr = &adjust,
```

Comment:
I see nothing that's pipeline-specific here, I really don't understand what 'adjust' means here. If the message destination is the base firmware module how would one change pipeline Y ?

Reply:
IIUC it doesn't matter which specific pipeline requests the change in clock rate, we just add up all pipelines and configure the clock with the sum